src/transport_ofi.c

```diff
@@ -271,6 +311,181 @@ typedef enum{
+static inline
+void shmem_transport_ofi_stx_rand_init() {
+    rand_pool_indices = malloc(shmem_transport_ofi_stx_max*sizeof(uint32_t));
+    if (rand_pool_indices == NULL) {
+        RAISE_ERROR_STR("Error: out of memory when allocating OFI random STX indices");
+    }
+
+    for(int i=0; i<shmem_transport_ofi_stx_max; i++) {
+        rand_pool_indices[i] = i;
+    }
+    /* Start at the last element: */
+    rand_pool_top_idx = shmem_transport_ofi_stx_max-1;
+
+    return;
+}
+
+static inline
+int fair_rand() {
+    /* Choose an STX index from the unselected subset */
+    int choice = rand() % rand_pool_top_idx;
+
+    /* Swap the value at the chosen index with the value at the top index */
+    int tmp = rand_pool_indices[choice];
+    rand_pool_indices[choice] = rand_pool_indices[rand_pool_top_idx];
+    rand_pool_indices[rand_pool_top_idx] = tmp;
+
+    /* If we've reached the first element, then start over from the top
+     * (skipping the 0th element reserved for the default context) */
+    if (rand_pool_top_idx == 1) {
+        rand_pool_top_idx = shmem_transport_ofi_stx_max - 1;
+    } else {
+        rand_pool_top_idx--;
+    }
+
+    return tmp;
+}
+
+void shmem_transport_ofi_stx_rand_fini() {
+    free(rand_pool_indices);
+    return;
+}
+
+#ifdef ENABLE_THREADS
+static inline
+void shmem_transport_ofi_stx_pool_search(shmem_transport_ctx_t *ctx)
+{
+    int i, count = 0;
+    switch (shmem_transport_ofi_stx_share_alg) {
+        case (ROUNDROBIN):
+            i = stx_start_idx;
+            while (count < shmem_transport_ofi_stx_max) {
+                if( shmem_transport_ofi_stx_pool[i].ref_cnt == 0 &&
+                   !shmem_transport_ofi_stx_pool[i].is_private ) {
+                    ctx->stx_idx = i;
+                    stx_start_idx = i;
+                    break;
+                } else {
+                    i = (i + 1) % shmem_transport_ofi_stx_max;
+                    count++;
+                }
+            }
+
+            if (count == shmem_transport_ofi_stx_max) {
+                /* No free STX is available, so share with shmem_ctx_default */
+                ctx->stx_idx = 0;
+                stx_start_idx = 1;
+                if (ctx->options & SHMEM_CTX_PRIVATE) {
+                    RAISE_WARN_STR("No OFI STX available, so sharing with the default context.");
+                }
+            }
+            break;
+
+        case (RANDOM):
+            i = fair_rand();
+            while (count < shmem_transport_ofi_stx_max) {
+                if( shmem_transport_ofi_stx_pool[i].ref_cnt == 0 &&
+                   !shmem_transport_ofi_stx_pool[i].is_private ) {
+                    ctx->stx_idx = i;
+                    break;
+                } else {
+                    i = fair_rand();
+                    count++;
+                }
+            }
+            if (count == shmem_transport_ofi_stx_max) {
+                /* No free STX is available, so share with shmem_ctx_default */
```

Comment:
Once all of the STXs have been allocated, we could share with any STX that is shared.  You could store these in a second array and use the same algorithm (RR, Random, etc.) to select from the array of shared STXs.

But, let's see if we can merge this and then layer on the above afterward.